File: ice9/censor.py
from __future__ import annotations

from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .models import AnalysisResult

# Labels that warrant censoring by default. Deliberately excludes belly,
# armpits, feet, and faces — NudeNet detects all of these, but they are not
# content that most applications need to redact.
CENSOR_LABELS: frozenset[str] = frozenset({
    "FEMALE_BREAST_EXPOSED",
    "MALE_BREAST_EXPOSED",
    "FEMALE_GENITALIA_EXPOSED",
    "MALE_GENITALIA_EXPOSED",
    "BUTTOCKS_EXPOSED",
    "ANUS_EXPOSED",
})

# The API normalizes uploads to this maximum dimension before running workers.
# Bounding box coordinates are in that normalized space.
_API_MAX_DIMENSION = 1024


def censor(
    result: AnalysisResult,
    image: str | Path,
    *,
    method: str = "fill",
    labels: frozenset[str] | set[str] | None = None,
    min_confidence: float = 0.5,
    output: str | Path | None = None,
):
    """Draw censoring over nudenet detections on the original image.

    Args:
        result:         An AnalysisResult that includes nudenet output.
        image:          Path to the original image file.
        method:         One of 'fill' (black rectangle) or 'pixelate'.
        labels:         Set of nudenet labels to censor. Defaults to CENSOR_LABELS,
                        which covers genitalia, breasts, and buttocks. Pass a custom
                        set to expand or restrict what gets redacted.
        min_confidence: Minimum detection confidence to act on. Default 0.5.
        output:         If given, save the censored image to this path.

    Returns:
        A PIL Image with censored regions applied.

    Raises:
        Ice9Error:   nudenet results are not present on the result.
        ValueError:  Unknown method.
        ImportError: Pillow is not installed.
    """
    try:
        from PIL import Image, ImageDraw
    except ImportError:
        raise ImportError(
            "Pillow is required for image censoring: pip install Pillow"
        ) from None

    from .exceptions import Ice9Error

    if result.nudenet is None:
        raise Ice9Error(
            "nudenet results are not present — was nudenet included in the tier?"
        )

    if method not in ("fill", "pixelate"):
        raise ValueError(f"Unknown censor method {method!r}. Choose 'fill' or 'pixelate'.")

    effective_labels = labels if labels is not None else CENSOR_LABELS
    predictions = result.nudenet._data.get("predictions") or []

    img = Image.open(image).convert("RGB")
    orig_w, orig_h = img.size

    # The API resizes to _API_MAX_DIMENSION on the longest side before processing.
    # Scale bbox coordinates back to original image space.
    longest = max(orig_w, orig_h)
    if longest > _API_MAX_DIMENSION:
        bbox_scale = longest / _API_MAX_DIMENSION
    else:
        bbox_scale = 1.0

    for pred in predictions:
        if pred.get("label") not in effective_labels:
            continue
        if pred.get("confidence", 0) < min_confidence:
            continue

        raw = pred["bbox"]
        x1 = int(raw["x"] * bbox_scale)
        y1 = int(raw["y"] * bbox_scale)
        x2 = int((raw["x"] + raw["width"]) * bbox_scale)
        y2 = int((raw["y"] + raw["height"]) * bbox_scale)

        # Clamp to image bounds
        x1, y1 = max(0, x1), max(0, y1)
        x2, y2 = min(orig_w, x2), min(orig_h, y2)

        if x2 <= x1 or y2 <= y1:
            continue

        if method == "fill":
            draw = ImageDraw.Draw(img)
            draw.rectangle([x1, y1, x2, y2], fill=(0, 0, 0))

        elif method == "pixelate":
            region = img.crop((x1, y1, x2, y2))
            block = max(8, min(x2 - x1, y2 - y1) // 8)
            small = region.resize(
                (max(1, (x2 - x1) // block), max(1, (y2 - y1) // block)),
                Image.NEAREST,
            )
            pixelated = small.resize((x2 - x1, y2 - y1), Image.NEAREST)
            img.paste(pixelated, (x1, y1))

    if output is not None:
        img.save(output)

    return img

File: nodejs/src/censor.ts
import { readFile, writeFile } from "node:fs/promises";
import { extname } from "node:path";

import { Ice9Error } from "./errors.js";
import type { AnalysisResult, BBox, Prediction } from "./result.js";

export const CENSOR_LABELS = new Set([
  "FEMALE_BREAST_EXPOSED",
  "MALE_BREAST_EXPOSED",
  "FEMALE_GENITALIA_EXPOSED",
  "MALE_GENITALIA_EXPOSED",
  "BUTTOCKS_EXPOSED",
  "ANUS_EXPOSED",
]);

export type CensorMethod = "fill" | "pixelate";

export interface CensorOptions {
  method?: CensorMethod;
  labels?: Set<string> | string[];
  minConfidence?: number;
  output?: string;
  format?: "jpeg" | "png" | "webp";
}

async function loadSharp() {
  try {
    return (await import("sharp")).default;
  } catch (cause) {
    throw new Ice9Error(
      "Image censoring requires the optional 'sharp' package. Install it with: npm install sharp",
      { cause },
    );
  }
}

function outputFormat(
  requested: CensorOptions["format"],
  output: string | undefined,
  inputFormat: string | undefined,
): "jpeg" | "png" | "webp" {
  if (requested) return requested;
  const extension = output ? extname(output).toLowerCase() : "";
  if (extension === ".jpg" || extension === ".jpeg") return "jpeg";
  if (extension === ".png") return "png";
  if (extension === ".webp") return "webp";
  if (output && extension) {
    throw new TypeError(
      `Cannot infer a supported censor output format from ${JSON.stringify(extension)}. ` +
      "Use a .jpg, .png, or .webp filename, or set options.format.",
    );
  }
  return inputFormat === "png" || inputFormat === "webp" ? inputFormat : "jpeg";
}

function clampBox(bbox: BBox, width: number, height: number) {
  const x1 = Math.max(0, Math.trunc(bbox.x));
  const y1 = Math.max(0, Math.trunc(bbox.y));
  const x2 = Math.min(width, Math.trunc(bbox.x + bbox.width));
  const y2 = Math.min(height, Math.trunc(bbox.y + bbox.height));
  return { x1, y1, x2, y2, width: x2 - x1, height: y2 - y1 };
}

function shouldCensor(
  prediction: Prediction,
  labels: Set<string>,
  minConfidence: number,
): prediction is Prediction & { bbox: BBox } {
  return Boolean(
    prediction.label &&
      prediction.bbox &&
      labels.has(prediction.label) &&
      (prediction.confidence ?? 0) >= minConfidence,
  );
}

export async function censor(
  result: AnalysisResult,
  image: string | Buffer | Uint8Array,
  options: CensorOptions = {},
): Promise<Buffer> {
  const sharp = await loadSharp();
  if (!result.nudenet) {
    throw new Ice9Error(
      "nudenet results are not present — was nudenet included in the tier?",
    );
  }

  const method = options.method ?? "fill";
  if (method !== "fill" && method !== "pixelate") {
    throw new TypeError(
      `Unknown censor method ${JSON.stringify(method)}. Choose 'fill' or 'pixelate'.`,
    );
  }

  const labels =
    options.labels instanceof Set
      ? options.labels
      : new Set(options.labels ?? Array.from(CENSOR_LABELS));
  const minConfidence = options.minConfidence ?? 0.5;
  const input =
    typeof image === "string" ? await readFile(image) : Buffer.from(image);

  const base = sharp(input).ensureAlpha();
  const meta = await base.metadata();
  const width = meta.width ?? 0;
  const height = meta.height ?? 0;
  let current = base;

  for (const prediction of result.nudenet.predictions) {
    if (!shouldCensor(prediction, labels, minConfidence)) {
      continue;
    }

    const box = clampBox(prediction.bbox, width, height);
    if (box.width <= 0 || box.height <= 0) {
      continue;
    }

    if (method === "fill") {
      const overlay = await sharp({
        create: {
          width: box.width,
          height: box.height,
          channels: 4,
          background: { r: 0, g: 0, b: 0, alpha: 1 },
        },
      })
        .png()
        .toBuffer();

      current = current.composite([{ input: overlay, left: box.x1, top: box.y1 }]);
      continue;
    }

    const region = await current
      .clone()
      .extract({ left: box.x1, top: box.y1, width: box.width, height: box.height })
      .toBuffer();
    const block = Math.max(8, Math.floor(Math.min(box.width, box.height) / 8));
    const pixelated = await sharp(region)
      .resize(Math.max(1, Math.floor(box.width / block)), Math.max(1, Math.floor(box.height / block)), {
        kernel: "nearest",
      })
      .resize(box.width, box.height, { kernel: "nearest" })
      .png()
      .toBuffer();

    current = current.composite([{ input: pixelated, left: box.x1, top: box.y1 }]);
  }

  const format = outputFormat(options.format, options.output, meta.format);
  const output = await (format === "png"
    ? current.png()
    : format === "webp"
      ? current.webp()
      : current.jpeg()
  ).toBuffer();
  if (options.output) {
    await writeFile(options.output, output);
  }
  return output;
}
